src/nvim/drawline.c

```diff
@@ -391,21 +392,34 @@ static bool use_cursor_line_nr(win_T *wp, linenr_T lnum, int row, int startrow,
 static inline void get_line_number_str(win_T *wp, linenr_T lnum, char_u *buf, size_t buf_len)
 {
   long num;
+  long rnum = -1;
   char *fmt = "%*ld ";
 
   if (wp->w_p_nu && !wp->w_p_rnu) {
     // 'number' + 'norelativenumber'
     num = (long)lnum;
   } else {
     // 'relativenumber', don't use negative numbers
-    num = labs((long)get_cursor_rel_lnum(wp, lnum));
+    rnum = num = labs((long)get_cursor_rel_lnum(wp, lnum));
     if (num == 0 && wp->w_p_nu && wp->w_p_rnu) {
       // 'number' + 'relativenumber'
       num = lnum;
       fmt = "%-*ld ";
     }
   }
 
+  if (*wp->w_p_nuf != NUL) {
+    // 'numberformat', set variables
+    typval_T tv = { .v_type = VAR_NUMBER };
+
+    num = ' ';
+    fmt = "%*c ";
+    tv.vval.v_number = wp->w_p_nu ? lnum : 0;
+    set_var(S_LEN("g:nuflnum"), &tv, false);
+    tv.vval.v_number = rnum;
+    set_var(S_LEN("g:nufrnum"), &tv, false);
```

Comment:
Set global variables to be used in the format expression eval block `%{}`. We could also add new format expression items containing the to be printed (relative)number. To do so we would have to pass along these values to `build_stl_str_hl()` or store them in the window structure or make them an internal global variable instead. Should allow for better performance for more simple cases that then don't require an eval block. More complicated cases need to be evaluated inside an eval block anyways though.

Reply:
> We could also add new format expression items containing the to be printed (relative)number.

Rather not. Let's optimize `v:` access instead (if it's not already optimized).